[Infra] add services to host detail ui

neptunian opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/168641.  

- Adds a new UI component to the host detail in a new collapsible, 'Services'. On both the flyout and asset details page when viewing an asset of type host only, this component should exist and list related services to the host.
- Removes "APM Services" link that used to exist in the detail view

<img width="1235" alt="Screenshot 2024-02-09 at 12 26 15 PM" src="https://github.com/elastic/kibana/assets/1676003/1a108d5b-437b-4dfd-bd04-9f137abb40d7">


##  Manual Testing

### Host with services
- Go to the host view and find a host with services using the edge cluster, by opening the flyout of various hosts
- Clicking on "show all" should take you to APM services inventory filtered by the host and showing the same services with time range persisted
- Clicking on an individual service from the flyout should take you to the APM service overview with timerange persisted
- Same should work on the node details

### Host with no services
<img width="1240" alt="Screenshot 2024-02-09 at 6 02 12 PM" src="https://github.com/elastic/kibana/assets/1676003/a3a32d2a-cd42-46c2-bd6a-79a11973ba04">

- Go to the host view and find a host with no services using the edge cluster 
- Open the host flyout and there should be a message that no services were found
- Same should work on the node details


### APM Services link removed 
The link seen below (top right) should not exist
<img width="1256" alt="Screenshot 2024-02-09 at 5 51 11 PM" src="https://github.com/elastic/kibana/assets/1676003/f79068ad-3f42-4471-a6e1-4e4bb8fe2819">

Commit:
add services to host detail ui

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

Commit:
translation fix

Commit:
include system metrics services

neptunian commented on x-pack/plugins/infra/server/lib/host_details/get_services.ts:
```diff
@@ -79,12 +79,35 @@ export const getServices = async (
           {
             term: {
-              'metricset.name': 'transaction',
+              'processor.event': 'metric',
```
since we're only getting services by metrics in this query I thought i'd add this

neptunian commented on x-pack/plugins/infra/server/lib/host_details/get_services.ts:
```diff
@@ -79,12 +79,35 @@ export const getServices = async (
           {
-            term: {
-              'metricset.interval': '1m', // make this dynamic if we start returning time series data
+            bool: {
```
this clause is to also check for "app"  (system metrics). https://www.elastic.co/guide/en/observability/current/data-model-metrics.html .  some services have no transactions.

Commit:
fix lint error

neptunian committed:
Merge branch 'main' into 168641-services-to-host-detail-ui

Commit:
add functional test

Commit:
Merge branch '168641-services-to-host-detail-ui' of https://github.com/neptunian/kibana into 168641-services-to-host-detail-ui

neptunian committed:
Merge branch 'main' into 168641-services-to-host-detail-ui

neptunian committed:
Merge branch 'main' into 168641-services-to-host-detail-ui

elasticmachine commented:
Pinging @elastic/obs-ux-infra_services-team (Team:obs-ux-infra_services)

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

neptunian commented on x-pack/test/functional/apps/infra/hosts_view.ts:
```diff
@@ -209,16 +220,37 @@ export default ({ getPageObjects, getService }: FtrProviderContext) => {
             await pageObjects.header.waitUntilLoadingHasFinished();
             await pageObjects.assetDetails.overviewAlertsTitleExists();
           });
+
```
In hindsight I wish i'd created fresh host data instead of trying to keep this in the Overview tab describe.  Creating new host data would have required me to update other tests.  I'd like to go back and use the infra synthtrace client instead of the host data in esArchiver files to make it easier to update and create the test data.

neptunian commented on x-pack/test/functional/apps/infra/helpers.ts:
```diff
@@ -0,0 +1,48 @@
+
+const SERVICE_PREFIX = 'service';
+// generates traces, metrics for services
+export function generateAddServicesToExistingHost({
```
this depends on receiving a hardcoded host that i got from other test data.  would like to refactor to use synthtrace infra hosts with consistent host names.

neptunian commented on x-pack/test/functional/apps/infra/config.ts:
```diff
@@ -5,20 +5,83 @@
     testFiles: [require.resolve('.')],
+    services: {
+      ...services,
+      infraSynthtraceEsClient: (context: InheritedFtrProviderContext) => {
```
not using this client but will be and thought i would make it available now for others

Commit:
update tests

Commit:
Merge branch '168641-services-to-host-detail-ui' of https://github.com/neptunian/kibana into 168641-services-to-host-detail-ui

neptunian commented on x-pack/test/functional/apps/infra/config.ts:
```diff
@@ -5,20 +5,83 @@
+          refreshAfterIndex: true,
+        });
+      },
+      apmSynthtraceEsClient: async (context: InheritedFtrProviderContext) => {
```
add the apm synthrace es client

neptunian commented on x-pack/plugins/infra/public/components/asset_details/links/link_to_apm_services.tsx:
```diff
@@ -34,14 +34,16 @@ export const LinkToApmServices = ({ assetName, apmField }: LinkToApmServicesProp
     <RedirectAppLinks coreStart={services}>
       <EuiButtonEmpty
-        data-test-subj="infraAssetDetailsViewAPMServicesButton"
+        data-test-subj="infraAssetDetailsViewAPMShowAllServicesButton"
```
I repurposed this for the "Show All" link.  This used to be the "APM Services" link at the top right of page.

Commit:
fix type errors

Commit:
use our FtrProviderContext

neptunian committed:
Merge branch 'main' into 168641-services-to-host-detail-ui

Commit:
fix type inference

Commit:
fix type

Commit:
remove unused type

Commit:
remove unused type

Commit:
fix type

crespocarlos commented on x-pack/plugins/infra/public/components/asset_details/hooks/use_services.ts:
```diff
@@ -0,0 +1,58 @@
+  }, [makeRequest, request$]);
+
+  return {
+    error: error || null,
```
do we need this? `error` is already `InfraHttpError | null`

crespocarlos reviewed (approved):
LGTM

cauemarcondes commented:
I think the icons look a bit odd. They look too big. And the label, for example on the java service doesn't look centralized.

<img width="1726" alt="Screenshot 2024-02-12 at 09 42 11" src="https://github.com/elastic/kibana/assets/55978943/e6c1e556-3d2f-43bd-a598-9384aaeb4057">

Commit:
skip ff test in metrics UI to show still failing ci

neptunian committed:
Merge branch 'main' into 168641-services-to-host-detail-ui

Commit:
use synthtrace within the test

Commit:
Merge branch '168641-services-to-host-detail-ui' of https://github.com/neptunian/kibana into 168641-services-to-host-detail-ui

Commit:
undo adding service to infra config and use synthtrace in test

neptunian commented on x-pack/plugins/infra/public/components/asset_details/components/services_tooltip_content.tsx:
```diff
@@ -0,0 +1,41 @@
+import { FormattedMessage } from '@kbn/i18n-react';
+import { useLinkProps } from '@kbn/observability-shared-plugin/public';
+
+export const ServicesTooltipContent = React.memo(() => {
```
I don't think so.  I copied over the pattern others were using but will leave changing those out of this PR.  Thanks!

neptunian commented on x-pack/plugins/infra/public/components/asset_details/components/services_tooltip_content.tsx:
```diff
@@ -0,0 +1,41 @@
+    app: 'home',
+    hash: '/tutorial/apm',
+  });
+  const onClick = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
+    e.stopPropagation();
+  };
```
same ^^.

Commit:
fix link to apm tutorial reloading kibana

Commit:
small pr feedback changes

Commit:
improve appearance of the icons

neptunian commented:
> I think the icons look a bit odd. They look too big. And the label, for example on the java service, doesn't look centralized.
> 
> <img alt="Screenshot 2024-02-12 at 09 42 11" width="1726" src="https://private-user-images.githubusercontent.com/55978943/304011624-e6c1e556-3d2f-43bd-a598-9384aaeb4057.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NjU5MDksIm5iZiI6MTcwNzc2NTYwOSwicGF0aCI6Ii81NTk3ODk0My8zMDQwMTE2MjQtZTZjMWU1NTYtM2QyZi00M2JkLWE1OTgtOTM4NGFhZWI0MDU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjEyVDE5MjAwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzOTcxOWFhMDI2MTExZTM5YTk1YWJkODBmNzQxMjZmNWZjN2U2NDk2MWQ2YzZmMzk1NDc4NmZhZjNmNDUwNjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.65SwpNcY_OkaaLz8sCYkGbgkNAimeH20y6hcyn_5A5M">

Thanks @cauemarcondes . I think the java label looks off because the icon is rather tall.  AgentIcon's around kibana are almost all using the large size, but i do agree medium looks better.  Though I like many things tiny so not sure I'm a good judge.  I also added some padding so that the icon doesn't touch the edges of the badge.  In hindsight we probably shouldn't be using badges as links, like this, as this isn't their intended use, but we'll be removing them soon in favor of a table.

Same size icons with padding (large)
<img width="586" alt="Screenshot 2024-02-12 at 2 35 19 PM" src="https://github.com/elastic/kibana/assets/1676003/ed6a5e85-8605-4ac7-ac4e-b3c9335f3408">

Medium size icons with padding
<img width="546" alt="Screenshot 2024-02-12 at 2 35 45 PM" src="https://github.com/elastic/kibana/assets/1676003/99295235-b8ef-4008-a03a-661fbf2d60e0">

@roshan-elastic do you have a preference?  I've switched it to medium size for now.

Commit:
remove skipping ff test tag

neptunian committed:
Merge branch 'main' into 168641-services-to-host-detail-ui

achyutjhunjhunwala commented on x-pack/test/functional/apps/infra/hosts_view.ts:
```diff
@@ -127,8 +148,29 @@ export default ({ getPageObjects, getService }: FtrProviderContext) => {
+      });
+      const kibanaVersion = await kibanaClient.fetchLatestApmPackageVersion();
+      await kibanaClient.installApmPackage(kibanaVersion);
+      synthtraceApmClient = new ApmSynthtraceEsClient({
```
Rather than creating the client inside a test file, you can create the client at FTR Context level and access it with `getService`. This way all the test files can access synthtrace easily.

For reference on how you can do this - https://github.com/elastic/kibana/blob/65291c64932389290cdf3a81017b3cd4f54bd143/x-pack/test/functional/apps/observability_logs_explorer/config.ts

neptunian replied:
I decided not to do that because it causes our [Firefox tests](https://github.com/elastic/kibana/blob/main/x-pack/test/functional/config.firefox.js) due to it not having access to the app config.

See https://github.com/elastic/kibana/issues/175064.  I'm planning to followup creating a shared service.

achyutjhunjhunwala replied:
I just saw the the broken types in Log Explorer. It's a good catch, i will push a fix for Log Explorer soon.

Regarding creating it as a shared service, that could be a really good idea but that could also be expensive due to it installing APM Package Integration on initialisation. 

When i extended it to Log Explorer, i found that it was being extended for Infra and other projects and thought of moving it to common but then i was not sure how bad that would impact the performance. But happy to see it getting some more love and being extended to common.

achyutjhunjhunwala replied:
Regarding FF test, my understanding was Appex team is slowing deprecating FF tests.

neptunian replied:
Yea, I'm thinking there might need to be a work around for installing the APM integration.  There is some chat happening about it I'll send you.

Seems like the appex team is letting plugin owners decide whether to be in FF or not.  I did talk to Felix and he said it would probably be fine for us to remove it (affects logs ui).  I didn't necessarily want to remove it because of this issue, though, as appex said we probably should not be fetching that integration in a service anyway.

neptunian committed:
Merge branch 'main' into 168641-services-to-host-detail-ui

neptunian committed:
Merge branch 'main' into 168641-services-to-host-detail-ui

roshan-elastic commented:
Look brilliant @neptunian - nice write up...very clear!

neptunian committed:
Merge branch 'main' into 168641-services-to-host-detail-ui

cauemarcondes reviewed (approved):
LGTM

roshan-elastic commented:
> @roshan-elastic do you have a preference? I've switched it to medium size for now.

Sorry @neptunian, missed this. Happy with medium.
